docs/module-caching.md

```diff
@@ -0,0 +1,178 @@
+
+## Invariants
+
+1. Importing a module with the same string specifier twice from the same script or module results in a single evaluation (this is already guaranteed by ECMA-262):
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   import { check as c2 } from "./file.js";
+   assert(c1 === c2);
+
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   const { check: c2 } = await import("./file.js");
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   const { check: c1 } = await import("./file.js");
+   const { check: c2 } = await import("./file.js");
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+2. Importing a module with the same string specifier from the top-level or from any module blocks defined in the same source text module and evaluated the same Realm results in a single evaluation:
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   const mod = module { export { check } from "./file.js"; };
+   const { check: c2 } = await import(mod);
+   assert(c1 === c2);
+
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   const mod1 = module { export { check } from "./file.js"; };
+   const mod2 = module { export { check } from "./file.js"; };
+   const { check: c2 } = await import(mod1);
+   const { check: c2 } = await import(mod2);
+   assert(c2 === c2);
+
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   import { check as c1 } from "./file.js";
+   const mod = module {
+      const nested = module { export { check } from "./file.js"; };
+      export const { check } = await import(nested);
+   };
+   const { check: c2 } = await import(mod);
+   assert(c1 === c2);
+
+   // file.js
+   export const check = {};
+   ```
+3. Importing the same module block twice _from the same Realm_, even if from different scripts or modules, results in a single evaluation:
+   ```js
+   globalThis.count = 0;
+   const mod = module { globalThis.count++; };
+   await import(mod);
+   await import(mod);
+   assert(globalThis.count === 1);
+   ```
+   ```js
+   // main.js
+   import { mod, check as c1 } from "./dep.js";
+   const { check: c2 } = await import(mod);
+   assert(c1 === c2);
+
+   // dep.js
+   export const mod = module { export const check = {}; };
+   export const { check } = await import(mod);
+   ```
+4. A module block is evaluated in the same realm as where it is imported:
+   ```js
+   const mod = module { globalThis.modEvaluated = true; };
+   const realm = createLegacyRealm(); // [1]
+   await realm.eval(`s => import(s)`)(mod);
```

Comment:
this one is very interesting. since `mod instanceof Module`, what happen if I do `import(mod)` before or after `realm.eval()`? To be more clear, this is the question:

 * `import(mod) !== realm.eval('s => import(s)')(mod)`

Reply:
To be clear, the inconsistency here with my mental model might be coming from the fact that you might do `new Module(moduleSource)` twice, and it produces two different modules.

Reply:
I wrote this document a while ago and I haven't checked if it's still correct. The proposal initially cached modules based on the realm of the import call instead of the one of the module expression.

This example is probably just outdated, I have to go through the document and review it. When in doubt, please check the proposal spec text!